### ISSUE TITLE
Enhancement: Add "Ship to all countries, except for..."

### ISSUE THREAD

**Issue by KoolPal** [woocommerce/woocommerce#17667]
**Objective**: Under Shipping locations, get another drop down "Ship to all countries, except for..."

This would help help manage multiple shipping countries which need to be excluded, for e.g. Iran, North Korea, Iraq, etc,

**Current option available**: You need to select "Ship to specific countries only", Choose all countries and then manually remove the countries you DO NOT WANT TO SHIP TO.

This is too cumbersome and unmanageable if you wish to review this excluded countries list from time to time.

This was originally requested in 2013 via ""Allowed Countries" setting is restricting billing address (not just shipping address) [#3215](https://github.com/woocommerce/woocommerce/issues/3215)" & "Add Separate Allowed Countries for Billing / Shipping [#3255](https://github.com/woocommerce/woocommerce/issues/3255)"
This was resolved since the requesters wanted this limitation of shipping only to 1 country.

In 2016, "Restrict Specific Countries [#10436](https://github.com/woocommerce/woocommerce/issues/10436)" was raised for simplifying the country selection.
Somehow the idea got diluted by the time it was implemented to only Selling countries via "Enhancement add countries exceptions billing and shipping [#10750](https://github.com/woocommerce/woocommerce/pull/10750)"

Can we extend this option "Ship to all countries, except for..." for Shipping countries as well?

In case I have missed some obvious setting than the current cumbersome option mentioned above, I would love to understand that!

Thanks!

**Comment by mikejolley** [comment]
If you narrow it down in `Selling countries` first, you can make shipping locations set to `ship to all countries you sell to` so they share the same options and exclusions.

Enhancing `Shipping location` is probably not worth it at this point because it overlaps quite heavily with Shipping Zones. If you don't have a zone for a location, no shipping is offered there anyway. I can see this setting being removed in favour of zones in the future, after review, now zones have matured.

Thanks

**Comment by KoolPal** [comment]
@mikejolley Perhaps I should have added the use case: Allow the customer with billing address from any country, but only allow the order to be shipped to the allowed countries excluding Iran, North Korea, Iraq, etc (just as an example to maintain context).

If I were to implement your suggestion in limiting the countries in Selling countries, this prevents billing address from those countries as well. In this case, customers with billing address in Iran, North Korea, Iraq, etc cannot buy from the store at all.

With zones use are you suggesting Sell to all countries but remove Iran, North Korea, Iraq, etc.? If yes, when a customer enters either of these countries as a shipping address it will show the dreaded “No shipping method has been selected. Please double check your address, or contact us if you need any help.” error which will lead to additional confusion.

In short we welcome customers to buy from anywhere, but limit our shipping to exclude certain countries.

As I pointed out above, this is possible right now, but too cumbersome.

In case you are unwilling to consider this request, would you suggest some solution to this particular use case?

Thanks.

**Comment by mikejolley** [comment]
Surely you'll have zones for specific countries to reflect different prices? You can setup an Iraq zone for example, with no shipping methods, to electively exclude it.

**Comment by KoolPal** [comment]
@mikejolley Would your suggestion of `"setting up an Iraq zone for example, with no shipping methods"` not lead to `“No shipping method has been selected. Please double check your address, or contact us if you need any help.”` error which will lead to additional confusion?

**Comment by mikejolley** [comment]
"There are no shipping methods available. Please ensure that your address has been entered correctly, or contact us if you need any help."

**Comment by KoolPal** [comment]
@mikejolley Thank you for correcting the error message, but the moot point remains,

Seeing your reluctance to make this change in the core, would you be kind enough to assist with some code snippet to get this functionality in my store?

Code snippet to add another drop down under Shipping locations,  "Ship to all countries, except for..."

This help would be highly appreciated.

**Comment by mikejolley** [comment]
Why does the point remain? Zones solve this - the ship to countries option is not even needed if you use zones.

**Comment by KoolPal** [comment]
@mikejolley Sorry but either I am missing the obvious OR I am unable to explain the issue clearly.

I shall try to explain clearly again.

If you help in providing a Code snippet to add another drop down under Shipping locations, "Ship to all countries, except for..." and I mark **Iraq** in this country.

Now when a customer chooses a Shipping country as Iraq, he will get the notice stating: 
"Unfortunately **we do not ship to IQ**. Please enter an alternative shipping address."
This is the **desired result**. The customer will know EXACTLY what is the issue and can take corrective action.

If I use the zones solution as suggested by you, he will get the notice stating:
`"There are no shipping methods available. Please ensure that your address has been entered correctly, or contact us if you need any help."`
This will be **confusing to the customer** and an undesired result. He will not know what to do next. He will probably end up writing to us **_OR worse abandon the cart_**.

I hope I have been able to explain better with this example.

Having said all this, I repeat, would you be kind enough to assist with **some code snippet** to get this functionality in my store?

Code snippet to add another drop down under Shipping locations, "Ship to all countries, except for..."

Thank you for your patience.

**Comment by mikejolley** [comment]
Why don't we update the wording "There are no shipping methods available. Please ensure that your address has been entered correctly, or contact us if you need any help."?

I'm hesitant to add more options for the ship to setting because I think that option could potentially be culled in favour of purely zones, pending feedback of course.

**Comment by KoolPal** [comment]
@mikejolley No matter how it is worded, the customer will not get a specific error message _**which is that we do not ship to that country.**_
 
I understand your hesitation in making this change to the core, hence I am **requesting an interim code snippet** which can used while this point is being discussed / evaluated.

Currently, my workaround is very messy as can be seen from the screenshot below. Plus keeping track of excluded countries needs to be done offline which is prone to errors.

![image](https://user-images.githubusercontent.com/454883/32837832-fa01cb1e-ca34-11e7-894d-465a55a7b4ce.png)

**Comment by mikejolley** [comment]
Sorry but we don't offer code snippets/customisations here...

> No matter how it is worded, the customer will not get a specific error message which is that we do not ship to that country.

Which they do with this other field anyway!

**Comment by willgorham** [comment]
Maybe something as simple as adjusting the notice to "There are no shipping methods are available **for this location**. Please ensure that your address has been entered correctly, or contact us if you need any help." would suffice? Or even "We do not ship to this location. Please ensure that your address has been entered correctly, or contact us if you need any help." I agree that the current "No shipping methods are available" phrasing isn't very intuitive to a customer.

**Comment by mikejolley** [comment]
@willgorham I think it could also mention the location/country - like the other notice.

**Comment by KoolPal** [comment]
Jeez, I didn't think that English was so hard for Woocommerce team to comprehend

**Comment by edadams** [comment]
I've been directed here by @willgorham from #20117, from which I learned that there is not currently a way to customize the message, `"There are no shipping methods available. Please ensure that your address has been entered correctly, or contact us if you need any help."`. I'll throw my :+1: vote on this because that message is quite vague (and although I did find a workaround to change it, it's not really ideal).

**Comment by KoolPal** [comment]
@mikejolley, I was pleasantly surprised to see this option in General Settings in version 3.5.3. 

When did this get released? This issue is closed without any reference to code changes. 

Changelog is silent on this as well. Any reason this was change was introduced in stealth mode?

Or I am confused if I found some code snippet which fixed this. I do not recall now. I am stuck with other challenges within Woocommerce at the moment.

![image](https://user-images.githubusercontent.com/454883/51299728-6f169400-1a4f-11e9-8463-9e3f9e86afb4.png)

**Comment by mikejolley** [comment]
Thats 'sell to', not 'ship to'

**Comment by KoolPal** [comment]
Ah OK. I have a bad cold since a few days. Error between keyboard and chair 😀
